server/grpc_service.go

```diff
@@ -502,37 +508,39 @@ func (s *GrpcServer) forwardTSO(stream pdpb.PD_TsoServer) error {
 }
 
 func (s *GrpcServer) forwardTSORequestWithDeadLine(
-	ctx context.Context, request *pdpb.TsoRequest, forwardStream tsopb.TSO_TsoClient,
+	forwardCtx context.Context,
+	cancelForward context.CancelFunc,
+	forwardStream tsopb.TSO_TsoClient,
+	request *pdpb.TsoRequest,
+	tsDeadlineCh chan<- *tsoutil.TSDeadline,
 ) (*tsopb.TsoResponse, error) {
-	defer logutil.LogPanic()
-	// Create a context with deadline for forwarding TSO request to TSO service.
-	ctxTimeout, cancel := context.WithTimeout(ctx, tsoutil.DefaultTSOProxyTimeout)
-	defer cancel()
-
-	tsoProxyBatchSize.Observe(float64(request.GetCount()))
+	done := make(chan struct{})
+	dl := tsoutil.NewTSDeadline(tsoutil.DefaultTSOProxyTimeout, done, cancelForward)
+	select {
+	case tsDeadlineCh <- dl:
```

Comment:
It seems only check whether timeout is in L520, if it is stuck in L526, what will we do?